### PR TITLE
feat(hints): add NewHint#47

### PR DESCRIPTION
**Pull request by MegaRedHand**
## Description

This PR adds the missing hint _NewHint#47_, now known as `UINT512_UNSIGNED_DIV_REM`.

In addition to that:
 - splits tests and libraries for the `uint384` and `uint384_extension` cairo contracts, to reuse them in other contracts.
 - unifies pack and split into a generic function in `uint_utils.rs`, and changes `UintNNN` specific pack and split to use the generic ones.
 - specifies full import path in some macros, and fix resulting "unused import" errors.
 - add Uint512 type (maybe this types could be implemented with a generic or macro?)

### CODE REVIEW

File: src/hint_processor/builtin_hint_processor/fq.rs
Comment (Oppen) [inline src/hint_processor/builtin_hint_processor/fq.rs:1]: What does "fq" stand for?

Comment (MegaRedHand) [reply src/hint_processor/builtin_hint_processor/fq.rs:1]: AFAIK "fq" comes from fq, the symbol for a finite field of _q_ elements

Comment (Oppen) [reply src/hint_processor/builtin_hint_processor/fq.rs:1]: I'd love a comment at the top of the file for context, but it's non-blocking.

Comment (MegaRedHand) [reply src/hint_processor/builtin_hint_processor/fq.rs:1]: Done. Also renamed `fq.cairo` -> `fq_test.cairo` and `uint512.cairo` -> `fq.cairo` to reduce confusion, as that was `fq.cairo`'s original name.

**Review by Oppen — APPROVED** [review]

File: src/hint_processor/builtin_hint_processor/uint256_utils.rs [hunk 1]
```diff
@@ -82,11 +89,11 @@ impl<'a> From<Felt252> for Uint256<'a> {
     }
 }
 
-pub(crate) fn pack(num: Uint256) -> BigUint {
+pub(crate) fn u256_pack(num: Uint256) -> BigUint {
     (num.high.to_biguint() << 128) + num.low.to_biguint()
 }
 
-pub(crate) fn split(num: &BigUint) -> [Felt252; 2] {
+pub(crate) fn u256_split(num: &BigUint) -> [Felt252; 2] {
```
Comment (fmoletta) [inline src/hint_processor/builtin_hint_processor/uint256_utils.rs:96]: nit: Could these be Uint256 methods instead?

File: src/hint_processor/builtin_hint_processor/uint384.rs [hunk 2]
```diff
@@ -62,23 +63,12 @@ impl Uint384ExpandReduced<'_> {
...
-pub(crate) fn pack(num: BigInt3, num_bits_shift: usize) -> BigUint {
-    let limbs = [num.d0, num.d1, num.d2];
-    #[allow(deprecated)]
-    limbs
-        .into_iter()
-        .enumerate()
-        .map(|(idx, value)| value.to_biguint().shl(idx * num_bits_shift))
-        .sum()
+pub(crate) fn u384_split(num: &BigUint) -> [Felt252; 3] {
+    split(num, 128)
```
Comment (fmoletta) [inline src/hint_processor/builtin_hint_processor/uint384.rs:71]: Same as previous comment

File: src/hint_processor/builtin_hint_processor/uint256_utils.rs (hunk 1 above)
Comment (MegaRedHand) [reply src/hint_processor/builtin_hint_processor/uint256_utils.rs]: Done!

**Review by fmoletta — APPROVED** [review]

**Pull request merged**